AK/FloatingPointStringConversions.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022, David Tuin <davidot@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#ifdef KERNEL
+#    error This file should not be included in the KERNEL as it deals with doubles \
+           and there is no guraantee does not do any floating point computations.
+#endif
+
+#include <AK/StringView.h>
+
+namespace AK {
+
+static constexpr char floating_point_decimal_separator = '.';
+
+enum class FloatingPointError {
+    None,
+    NoOrInvalidInput,
+    OutOfRange,
+    RoundedDownToZero
+};
+
+template<FloatingPoint T>
+struct FloatingPointParseResults {
+    char const* end_ptr { nullptr };
+    FloatingPointError error = FloatingPointError::None;
+    T value {};
+
+    [[nodiscard]] bool parsed_value() const
+    {
+        // All other errors do indicate out of range but did produce a valid value.
+        return error != FloatingPointError::NoOrInvalidInput;
+    }
+};
+
+/// This function finds the first floating point within [start, end). The accepted format is
+/// intentionally as lenient as possible. If your format is stricter you must validate it
+/// first. The format accepts:
+/// - An optional sign, both + and - are supported
+/// - 0 or more decimal digits, with leading zeros allowed [1]
+/// - A decimal point '.', which can have no digits after it
+/// - 0 or more decimal digits, unless the first digits [1] doesn't have any digits,
+///   then this must have at least one
+/// - An exponent 'e' or 'E' followed by an optional sign '+' or '-' and at least one digit
+/// This function additionally detects out of range values which have been rounded to
+/// [-]infinity or 0 and gives the next character to read after the floating point.
+template<FloatingPoint T = double>
+FloatingPointParseResults<T> parse_first_floating_point(char const* start, char const* end);
```

Comment:
Not sure what the implications would be, but can we change this to take a `StringView`? Seems a bit strange to add a new AK function using c strings in this day and age. I'm sure LibC could wrap its input in one...?

Reply:
I made these take `char*` because it's the simplest interface.
We can't just use a StringView in libc, since we would have to do `strlen` and that makes `strtod` regress in performace as you can see above.

I think that with the convenience functions on the String types you barely want to call these functions anyway. If you feel this should be changed we can except for the one not taking and `end`